xmtp_mls/src/lib.rs

```diff
@@ -20,12 +20,81 @@ pub mod verified_key_package_v2;
 mod xmtp_openmls_provider;
 
 pub use client::{Client, Network};
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::sync::{Arc, LazyLock};
 use storage::{DuplicateItem, StorageError};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
 pub use xmtp_openmls_provider::XmtpOpenMlsProvider;
 
 pub use xmtp_id::InboxOwner;
 pub use xmtp_proto::api_client::trait_impls::*;
 
+/// A manager for group-specific semaphores
+#[derive(Debug)]
+pub struct GroupCommitLock {
+    // Storage for group-specific semaphores
+    locks: Mutex<HashMap<Vec<u8>, Arc<Semaphore>>>,
+}
+
+impl Default for GroupCommitLock {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+impl GroupCommitLock {
+    /// Create a new `GroupCommitLock`
+    pub fn new() -> Self {
+        Self {
+            locks: Mutex::new(HashMap::new()),
+        }
+    }
+
+    /// Get or create a semaphore for a specific group and acquire it, returning a guard
+    pub async fn get_lock_async(&self, group_id: Vec<u8>) -> Result<SemaphoreGuard, GroupError> {
+        let semaphore = {
+            let mut locks = self.locks.lock();
+            locks
+                .entry(group_id)
+                .or_insert_with(|| Arc::new(Semaphore::new(1)))
+                .clone()
+        };
+
+        let permit = semaphore.clone().acquire_owned().await?;
+        Ok(SemaphoreGuard {
+            _permit: permit,
+            _semaphore: semaphore,
+        })
+    }
+
+    /// Get or create a semaphore for a specific group and acquire it synchronously
+    pub fn get_lock_sync(&self, group_id: Vec<u8>) -> Result<SemaphoreGuard, GroupError> {
+        let semaphore = {
+            let mut locks = self.locks.lock();
+            locks
+                .entry(group_id)
+                .or_insert_with(|| Arc::new(Semaphore::new(1)))
+                .clone()
+        };
+
+        // Synchronously acquire the permit
+        let permit = semaphore.clone().try_acquire_owned()?;
```

Comment:
This is the part that I'm most uncertain about, it is most problematic in specific multi-threaded circumstances, which I think would be worth creating a test for, probably in `bindings_ffi` since that's where multi-threading is most relevant. wasm will always be single-threaded

- Create client
- create a reference to the client, and start a `stream_all_messages` with this reference in its own thread.
- clone two references to client
- spawn two threads each holding a reference to this client
- do a few syncs in each thread with each client reference. on the main thread, send a bunch of messages with a different user-client
- in addition to messages, can try to do operations on the group like updating the name/picture/etc, since those predominantly use the sync variant


the aim here is to try to create a situation that races the two syncs to acquire a (syncronous) permit. The worst case is if we lose messages in a stream, since it can be cumbersome for integrators to restart streams. An error from calling sync is less bad and recoverable by integrators, but we may want to create a descriptive error message for this case (like, "critical updates already ongoing" or something) so it doesn't cause too much turmoil when it happens. An error because a sync is happening when trying to update the group photo or something else is probably the least bad

Maybe can be done in a follow up?